### PR TITLE
podvm: Use local files for build and reference versions file

### PR DESCRIPTION
**Pull request by jtumber-ibm**
This introduces new Make targets in the root Makefile to create the podvm image.

If doing all from scratch `make docker-builder docker-binaries docker-image`. Else if binaries and builder already exist for current versions.yaml file then `make docker-image` will be enough.

_**I haven't update the rhel or centos versions yet**_

### CODE REVIEW

**Comment by stevenhorsman** [comment]
Sorry for the basic question - is this supposed to replace the current direct docker approach? If so I think we need to update the podvm README.md doc to reflect this?

**Comment by jtumber-ibm** [comment]
> Sorry for the basic question - is this supposed to replace the current direct docker approach? If so I think we need to update the podvm README.md doc to reflect this?

It is a helper rather than a complete replacement, but you're right the docs will need updating too.

**Comment by wainersm** [comment]
hi @jtumber-ibm ,  I hope to have some time today or tomorrow to review this. In meanwhile I was trying a different solution to allow caching kata-agent/AA to speed up the CI jobs, I am glad you sent this because I was having a really hard time to solve some cases :)

File: Makefile
Comment (katexochen) [inline Makefile:188]: We should remove the default version of Go in the docker image. Could this be also the right PR to also inject the other versions like rust etc?

**Review by katexochen — CHANGES_REQUESTED** [review]

File: Makefile
Comment (katexochen) [inline Makefile:208]:
```suggestion
docker-binaries:
	docker build -t $(PODVM_BINARIES_IMAGE) -f podvm/$(BINARIES_DOCKERFILE) \
	--build-arg BUILDER_IMG=$(PODVM_BUILDER_IMAGE) \
	--build-arg PODVM_DISTRO=$(PODVM_DISTRO) \
	--build-arg ARCH=$(ARCH) \
	--build-arg AA_KBC=$(AA_KBC) \
	$(DOCKER_OPTS) .

docker-image:
	docker build -t $(PODVM_IMAGE) -f podvm/$(PODVM_DOCKERFILE) \
	--build-arg BUILDER_IMG=$(PODVM_BUILDER_IMAGE) \
	--build-arg BINARIES_IMG=$(PODVM_BINARIES_IMAGE) \
	--build-arg PODVM_DISTRO=$(PODVM_DISTRO) \
	--build-arg ARCH=$(ARCH) \
	$(DOCKER_OPTS) .
```
There are some whitespace issues in your file. What editor are you using? :)

Comment (jtumber-ibm) [reply Makefile:208]: VS Code, although I recently switch workstations so quite a few settings were missed (I didn't have the settings sync enabled) :)

File: .dockerignore
Comment (jtumber-ibm) [reply .dockerignore:18]: It is annoying but in its current state it requires copying across the git directory into docker so we can get the commit hash for the `.git-commit` versioning in the top-level Makefile. 
We could refactor it to be able to parse in the commit calculated outside of docker.

File: Makefile
Comment (jtumber-ibm) [reply Makefile:188]: Yes to removing the default Go versions from the docker image so we don't have to keep that up to date.

Not sure what you mean by the other versions, they are also being included in the next lines, or something I've missed?

File: Makefile
Comment (katexochen) [reply Makefile]: I use the following in `settings.json` for normalizing whitespace:

```json
    "files.trimTrailingWhitespace": true,
    "files.trimFinalNewlines": true,
    "files.insertFinalNewline": true,
```

File: .dockerignore
Comment (katexochen) [reply .dockerignore:18]: My comment was just to remove the second newline here. ;)

Comment (jtumber-ibm) [reply .dockerignore:18]: Oh okay

**Comment by wainersm** [comment]
@jtumber-ibm how does this PR is related with https://github.com/confidential-containers/cloud-api-adaptor/pull/1018 ? does this supersedes PR 1018?

**Comment by wainersm** [comment]
@jtumber-ibm still trying to understand your changes, I just glanced at the code. One question that I have is about `f binaries and builder already exist for current versions.yaml file then make docker-image will be enough`...

Apart from `versions.yaml`,it seems that build flags and build scripts themselves that would invalidate a cache. For example, suppose we decide to change `ttrc` to false in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/podvm/Makefile.inc#L149 while keeping the same version of guest components.

File: Makefile
Comment (wainersm) [inline Makefile:175]: Missing `$(PODVM_DISTRO)` in binaries image name.

**Comment by jtumber-ibm** [comment]
> @jtumber-ibm how does this PR is related with #1018 ? does this supersedes PR 1018?

It does supersede #1018, I'll ask @yoheiueda if it is okay to close #1018

File: Makefile
Comment (jtumber-ibm) [reply Makefile:175]: Are the binaries DISTRO specific?

**Comment by jtumber-ibm** [comment]
> @jtumber-ibm still trying to understand your changes, I just glanced at the code. One question that I have is about `f binaries and builder already exist for current versions.yaml file then make docker-image will be enough`...
> 
> Apart from `versions.yaml`,it seems that build flags and build scripts themselves that would invalidate a cache. For example, suppose we decide to change `ttrc` to false in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/podvm/Makefile.inc#L149 while keeping the same version of guest components.

That is true, and a good point. I don't have a great answer for how to solve it though. Do you have a recommendation for how to track those changes / invalidate the cache when the build changes? My thought was to add a `build-version` entry into the versions.yaml but that isn't easy to track against the actual changes

File: Makefile
Comment (katexochen) [reply Makefile:175]: If they are dynamically linked they are, yes.

Comment (katexochen) [reply Makefile:175]: See https://github.com/confidential-containers/cloud-api-adaptor/issues/1093#issuecomment-1600991335 for an example.

**Comment by katexochen** [comment]
> > @jtumber-ibm still trying to understand your changes, I just glanced at the code. One question that I have is about `f binaries and builder already exist for current versions.yaml file then make docker-image will be enough`...
> > Apart from `versions.yaml`,it seems that build flags and build scripts themselves that would invalidate a cache. For example, suppose we decide to change `ttrc` to false in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/podvm/Makefile.inc#L149 while keeping the same version of guest components.
> 
> That is true, and a good point. I don't have a great answer for how to solve it though. Do you have a recommendation for how to track those changes / invalidate the cache when the build changes? My thought was to add a `build-version` entry into the versions.yaml but that isn't easy to track against the actual changes

I don't think we should build an entire build system from scratch. If we want such features, we will have to use a real build system like Nix or Bazel.

File: Makefile
Comment (jtumber-ibm) [reply Makefile:175]: Ah interesting, thanks. I'll make the name distro specific then

**Comment by wainersm** [comment]
> > > jtumber-ibm still trying to understand your changes, I just glanced at the code. One question that I have is about `f binaries and builder already exist for current versions.yaml file then make docker-image will be enough`...
> > > Apart from `versions.yaml`,it seems that build flags and build scripts themselves that would invalidate a cache. For example, suppose we decide to change `ttrc` to false in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/podvm/Makefile.inc#L149 while keeping the same version of guest components.
> > 
> > 
> > That is true, and a good point. I don't have a great answer for how to solve it though. Do you have a recommendation for how to track those changes / invalidate the cache when the build changes? My thought was to add a `build-version` entry into the versions.yaml but that isn't easy to track against the actual changes
> 
> I don't think we should build an entire build system from scratch. If we want such features, we will have to use a real build system like Nix or Bazel.

Yes, sure, don't want to re-invent the wheels. I am not sure about Nix, but even with bazel we would still have problems because the main concern is the build inside containers.

okay, back to the problem... ignore it? i.e. assume in a rare situation we might need to change build parameters but not `versions`, then defer the solution?

If not ignore then I can think of either:
 * Add a new property to `versions` to version the build configuration (e.g. `build_version=N`) that we can bump in case of we want to force the rebuild for whatever reason.
 * Apart from `versions.yaml`, compute the build files (Makefiles, etc) in the `VERSION_HASH`. This is the approach used by kata containers, btw.

Both approach are error prone: one might forget to update the `build_version`. One might forget to include new build files on the computation of the `VERSION_HASH`.

BTW, @fidencio has recently added (https://github.com/kata-containers/kata-containers/pull/8100)  the infra to allow building kata-agent separated from the entire rootfs and, IIUC, the idea is to cache the kata-agent binary just like the other components. In other words, soon we might be able to consume kata-agent binaries... In Fabiano we trust, In fabiano we rest our hopes!

**Comment by wainersm** [comment]
@jtumber-ibm perhaps one last thing not relate with code properly saying. When we (@bpradipt and I) wrote the first implementation, the idea of pulling the sources from git tree was to enable reproducibility. Switching to copying the local sources to the image we lose reproducibility if I am not mistaken. As a developer I love the simplicity that this solution brings to my life (I can even feel the inner peace on my heart growing...) but I'm not sure how it impacts on peer pods' distributors. Is this (loosing reproducibility) something already discussed on the community? I missed many discussions recently so it might be case....

**Comment by katexochen** [comment]
> When we (bpradipt and I) wrote the first implementation, the idea of pulling the sources from git tree was to enable reproducibility. 

IMO pulling code from GitHub has nothing to do with reproducibility. You can totally have reproducible builds from a local checkout. If you pull things from remote, you have to pin everything you pull in (by hash!). We never did that, and the way we pulled things, and still pull things, is not reproducible at all. Some examples from current main:

We don't even pin the base image by hash. The meaning of the tag changes multiple times a week:

https://github.com/confidential-containers/cloud-api-adaptor/blob/cf07e8adb388523a81c39483974fd2b8d0e03a13/podvm/Dockerfile.podvm_builder#L9


We pull CAA from main. If I build two times and you merge a PR in between, I get different results:

https://github.com/confidential-containers/cloud-api-adaptor/blob/cf07e8adb388523a81c39483974fd2b8d0e03a13/podvm/Dockerfile.podvm_builder#L57

Kata isn't pinned as well, as is any other of the many tools we are building:

https://github.com/confidential-containers/cloud-api-adaptor/blob/cf07e8adb388523a81c39483974fd2b8d0e03a13/podvm/Dockerfile.podvm_builder#L60

Currently, the only reproducible bit in our builds yq (and the stuff we already pinned in versions.yaml):

https://github.com/confidential-containers/cloud-api-adaptor/blob/cf07e8adb388523a81c39483974fd2b8d0e03a13/podvm/Dockerfile.podvm_builder#L14-L15

**Comment by katexochen** [comment]
> I am not sure about Nix, but even with bazel we would still have problems because the main concern is the build inside containers.

I don't get that point. You wouldn't use those tools within a container, they come with their own sandbox.

File: Makefile
Comment (jtumber-ibm) [reply Makefile:190]: I have moved the `yq` values out as well

File: Makefile.defaults
Comment (katexochen) [inline Makefile.defaults:12]: Can't we move this into `versions.yaml` ?

Comment (jtumber-ibm) [reply Makefile.defaults:12]: Do you think we should add a new section for remote files/binaries in the versions.yaml? i.e. At the moment the `tools` section doesn't have a field for checksums. Obviously YAML is flexible that we could just do that the `yq` ...

```yaml
tools:
  golang: 1.20.8
  rust: 1.72.0
  protoc: 3.11.4
  yq:
    version: ...
    checksum: ...
```

Although we don't have a strict schema to follow be nice to keep it consistent.

Comment (katexochen) [reply Makefile.defaults:12]: Your right. Let's skip it then. I think the versions.yaml thing won't be for every anyway, as it just can't replace a full build system.

**Review by katexochen — APPROVED** [review]
LGTM. Local build was successful, full build took 38 min.

**Review by stevenhorsman — COMMENTED** [review]
I've not had a chance to look through all the code, but I've run it locally with `ARCH=s390x make podvm-builder podvm-binaries podvm-image` and it created an image at least. I hit one problem on the way and I think we might need to update the documentation of how to build the podvm based on this as an easier way to run it locally. I hope to review it more fully next week. Thanks and sorry for the delays in the review

File: Makefile.defaults
Comment (katexochen) [reply Makefile.defaults:14]: One of those problems we could solve with a reproducible dev environment provided by Nix...

**Comment by surajssd** [comment]
What is blocking this PR?

**Comment by stevenhorsman** [comment]
> What is blocking this PR?

I think JT needs to rebase once they are back from vacation and it needs another review, which I intend to do once it's rebased to test the images that are created, and run it on s390x and x86, rather than just built a single one to test out the process. I think some doc updates might be important as well as I mention in my comment last week.

File: Makefile.defaults
Comment (jtumber-ibm) [reply Makefile.defaults:14]: https://github.com/mikefarah/yq/issues/1158#issuecomment-1081245199

We could specify `eval (e)` in our commands instead to ensure we are compatible with all the 4.x minor versions

**Comment by jtumber-ibm** [comment]
@surajssd @stevenhorsman rebased and doc update added

**Review by stevenhorsman — APPROVED** [review]
I recreated an s390x podvm based on the documentation and tried it out locally and it seems to pass the tests, so LGTM. Thanks for the updates!

**Pull request merged**
